discovery-provider/src/utils/redis_cache.py

```diff
@@ -0,0 +1,58 @@
+import redis
+import logging # pylint: disable=C0302
+import json
+import functools
+from src.utils.config import shared_config
+from flask.json import dumps
+from flask.globals import request
+logger = logging.getLogger(__name__)
+
+REDIS_URL = shared_config["redis"]["url"]
+REDIS = redis.Redis.from_url(url=REDIS_URL)
+
+# Redis Key Convention:
+# API_V1:path:queryparams:headers
+
+cache_prefix = "API_V1_ROUTE"
+# query params to always exclude from key construction
+exclude_param_set = {"app_name"}
+# headers to always include in key construction
+required_headers_set = {"X-User-ID"}
+default_ttl_sec = 60
+
+def extract_key():
+    path = request.path
+    req_args = request.args.items()
+    req_args = filter(lambda x: x[0] not in exclude_param_set, req_args)
+    req_args = sorted(req_args)
+    req_args = "&".join(["{}={}".format(x[0], x[1]) for x in req_args])
+    headers = []
+    for required_header in required_headers_set:
+        val = request.headers.get(required_header)
+        if val:
+            headers.append((required_header, val))
+    headers_str = "&".join(["{}={}".format(x[0], x[1]) for x in headers])
```

Comment:
Does this mean that we cache responses per user?
I don't think there's too much value in a per-user cache, I would say that we could ignore it if the header is present and only cache requests w/out the `x-user-id` header b/c it modified the result.

Reply:
Yeah good call-out, as per discussion on slack going to remove the header caching entirely here. Not quite sure why I added this.